src/Workspaces/Core/Portable/EmbeddedLanguages/DateTime/LanguageServices/DateTimePatternDetector.cs

```diff
@@ -0,0 +1,202 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using Microsoft.CodeAnalysis.EmbeddedLanguages.LanguageServices;
+using Microsoft.CodeAnalysis.LanguageServices;
+
+namespace Microsoft.CodeAnalysis.EmbeddedLanguages.DateTime.LanguageServices
+{
+    /// <summary>
+    /// Helper class to detect <see cref="System.DateTime"/> and <see cref="DateTimeOffset"/> format
+    /// strings in a document efficiently.
+    /// </summary>
+    internal sealed class DateTimePatternDetector
+    {
+        private const string FormatName = "format";
+
+        private static readonly ConditionalWeakTable<SemanticModel, DateTimePatternDetector?> _modelToDetector =
```

Comment:
I don't quite get what is going on here - will let someone more familiar with semantic model usage comment.